pkg/kv/kvserver/replica_init.go

```diff
@@ -20,77 +20,131 @@ import (
 	"go.etcd.io/raft/v3"
+	"go.etcd.io/raft/v3/raftpb"
 )
 
 const (
 	splitQueueThrottleDuration = 5 * time.Second
 	mergeQueueThrottleDuration = 5 * time.Second
 )
 
-// newReplica constructs a new Replica. If the desc is initialized, the store
-// must be present in it and the corresponding replica descriptor must have
-// replicaID as its ReplicaID.
-func newReplica(
-	ctx context.Context, desc *roachpb.RangeDescriptor, store *Store, replicaID roachpb.ReplicaID,
-) (*Replica, error) {
-	repl := newUnloadedReplica(ctx, desc.RangeID, store, replicaID)
-	repl.raftMu.Lock()
-	defer repl.raftMu.Unlock()
-	repl.mu.Lock()
-	defer repl.mu.Unlock()
-
-	// TODO(pavelkalinnikov): this path is taken only in tests. Remove it and
-	// assert desc.IsInitialized().
+// loadedReplicaState represents the state of a Replica loaded from storage, and
+// is used to initialize the in-memory Replica instance.
+type loadedReplicaState struct {
+	replicaID roachpb.ReplicaID
+	hardState raftpb.HardState
+	lastIndex uint64
+	replState kvserverpb.ReplicaState
+}
+
+// loadReplicaState loads the state necessary to create a Replica from storage.
+func loadReplicaState(
+	ctx context.Context,
+	eng storage.Reader,
+	desc *roachpb.RangeDescriptor,
+	replicaID roachpb.ReplicaID,
+) (loadedReplicaState, error) {
+	sl := stateloader.Make(desc.RangeID)
+	id, found, err := sl.LoadRaftReplicaID(ctx, eng)
+	if err != nil {
+		return loadedReplicaState{}, err
+	} else if !found {
+		return loadedReplicaState{}, errors.AssertionFailedf(
+			"r%d: RaftReplicaID not found", desc.RangeID)
+	} else if loaded := id.ReplicaID; loaded != replicaID {
+		return loadedReplicaState{}, errors.AssertionFailedf(
+			"r%d: loaded RaftReplicaID %d does not match %d", desc.RangeID, loaded, replicaID)
+	}
+
+	ls := loadedReplicaState{replicaID: replicaID}
+	if ls.hardState, err = sl.LoadHardState(ctx, eng); err != nil {
+		return loadedReplicaState{}, err
+	}
+	if ls.lastIndex, err = sl.LoadLastIndex(ctx, eng); err != nil {
+		return loadedReplicaState{}, err
+	}
+	if ls.replState, err = sl.Load(ctx, eng, desc); err != nil {
+		return loadedReplicaState{}, err
+	}
+	return ls, nil
+}
+
+// check makes sure that the replica invariants hold for the loaded state.
+func (r loadedReplicaState) check(storeID roachpb.StoreID) error {
+	desc := r.replState.Desc
+	if id := r.replicaID; id == 0 {
+		return errors.AssertionFailedf("r%d: replicaID is 0", desc.RangeID)
+	}
+
 	if !desc.IsInitialized() {
-		repl.assertStateRaftMuLockedReplicaMuRLocked(ctx, store.Engine())
-		return repl, nil
+		// An uninitialized replica must have an empty HardState.Commit at all
+		// times. Failure to maintain this invariant indicates corruption. And yet,
+		// we have observed this in the wild. See #40213.
+		if hs := r.hardState; hs.Commit != 0 {
+			return errors.AssertionFailedf(
+				"r%d/%d: non-zero HardState.Commit on uninitialized replica: %+v", desc.RangeID, r.replicaID, hs)
+		}
+		return nil
```

Comment:
TODO: assert `lastIndex` == 0?

Reply:
I think we want some method that asserts integrity of the state, but it should live in kvstorage since it's purely a function of the engine (well, two engines eventually) and this should also live closely to where we do the reconciliation, which again will be somewhere in storage before we even make in-mem replicas.

That said this shouldn't stop us from writing checks. I just think they should at some unspecified point in the future migrate over to `kvstorage` (and also be invoked from `LoadAndReconcileReplicas`).

Reply:
Agree here, this was my intention too. I'm thinking that `loadedStateReplica` should blend into your `kvstorage.Replica` from #95513 over time, so that loading and invariant checking is all there.